CIPs/0005.md

```diff
@@ -0,0 +1,74 @@
+
+## Goals
+
+- Reduce bandwidth requirements of the full nodes.
+- Maximize availability of the IBFT overlay network.
+
+## Proposed Solution
+
+The announce protocol has to be able to deal with two fundamental faulty scenarios.  For a given validator pair, when only one of the validators has the correct enodeURL of the other (the asymmetric ignorance case) and when both of the validators don't have the correct enodeURL of the other (the mutual ignorance case).
+
+To deal with these cases, this proposal proposes adding the following high level components to the existing Announce protocol.  1) Add the capability for a validator to directly communicate to another remote validator it's assigned enodeURL, 2) add fast discovery of validator enode mapping changes, and 3) add capability for a validator to query another validator for it's assigned enodeURL via a gossiped message.  Details of these components are in the subsections below.
+
+The asymmetric ignorance case can then be handled by having the validators directly communicate to each remote validator it's assigned enodeURL whenever the enodeURL mapping is updated.  Note that this will need to happen for all remote validators, even if the assigned URL is not updated for a subset of the validators, since all validators will need to know what is the updated version number.  Also, this direct communication will need to be handled when there is and is not an existing connection between validators.
+
+The mutual ignorance case can be handled by having each validator monitor when a validator enode mapping changes, and when it realizes that it's changed, it will gossip a query message to the validator for the updated assigned enodeURL.  That validator will then directly communicate it to the querying validator.
+
+### Direct communication of assigned enodeURL
+
+This component will include the following changes:
+
+1)  Announce message format will need to be changed so that only one authenticated and encrypted enodeURL (along with the version number) can be put in the message payload
+2)  On enodeURL mapping update, that validator will need to directly send an announce message to all validators. In the case of when the validator already has a connection with the remote validator, the originating validator should simply send an announce message to it.  In the case when that connection doesn't exist, then the announce message will need to be sent as part of the connection establishing handshake.
+
+Issue https://github.com/celo-org/celo-monorepo/issues/2667 tracks this component.
+
+### Fast discovery of changed enodeURL mapping
+
+This component will require all full nodes and validators to store a copy of a table that will contain the latest enodeURL mapping version number for each elected validator.  Whenver a validator updates it's enodeURL mapping, it will need to gossip a message containing the updated version number, and all nodes should update their table.
```

Comment:
Will non-validators be able to "spoof" these messages and trick full nodes into believing the version number is higher than it is?

Reply:
This message will contain a signature from the originating validator, so wouldn't be able to be spoofed.